xc/device_common.ml

```diff
@@ -338,22 +339,101 @@ let is_upstream_qemu domid =
     with_xs (fun xs -> xs.Xs.read (sprintf "/libxl/%d/dm-version" domid)) = "qemu_xen"
   with _ -> false
 
-let qmp_write_and_read domid ?(read_result=true) cmd  =
-  try
-    let open Qmp in
-    debug "QMP.Command for domid %d: %s" domid (string_of_message cmd);
-    let c = Qmp_protocol.connect (qmp_libxl_path domid) in
-    finally
-      (fun () ->
-         Qmp_protocol.negotiate c;
-         Qmp_protocol.write c cmd;
-         if read_result then
-           Some (Qmp_protocol.read c)
-         else None
-      )
-      (fun () -> Qmp_protocol.close c)
-  with e ->
-    error "Caught exception attempting to write qmp message: %s" (Printexc.to_string e);
-    None
+let qmp domid cmd  =
+  let open Qmp in
+  let cmd' = string_of_message cmd in
+  debug "QMP.Command for domid %d: %s" domid cmd';
+  let c = Qmp_protocol.connect (qmp_libxl_path domid) in
+  finally
+    (fun () ->
+       Qmp_protocol.negotiate c;
+       Qmp_protocol.write c cmd;
+       Qmp_protocol.read c
+    )
+    (fun () -> Qmp_protocol.close c)
+
+(** [make_id_generator] returns a function that creates unique IDs using
+ * an internal counter *)
+let make_id_generator name =
+  let count = ref 0 in
+  fun domid -> incr count; Printf.sprintf "%s-%06d-%d" name !count domid
+
+let make_qmp_id = make_id_generator "qmp"
+
+(** [qmp_send_cmd_internal connection omid cmd] sends [cmd] to QEMU, waits
+ * for the *matching* response, and returns the result. It uses an
```

Comment:
I think it would be useful to know why you would ever get an unmatched response. Do we send multiple commands on the same connection and then wait for replies? If so then ignoring the response means some other command never gets it reply. If not, is it just to protect against bugs in qemu?

Reply:
We should never be in a position where multiple commands are being sent on the same connection at the same time. However, QEMU broadcasts events to all connected monitors (i.e. QMP connections). So this is mainly to protect against unexpected events being received instead of the result of a command.

We use a separate monitor specifically to handle events so it is fine to ignore them on the command connection.

Reply:
Testing reveals that we receive EVENT messages while waiting for the result of a command:
```
Feb 16 16:04:19 localhost xenopsd-xc: [debug|localhost.localdomain|41 |Async.VBD.eject R:9a2c66e44a71|xenops] QMP.Command for domid 4: {"execute":"eject","id":"qmp-000011-4","arguments":{"device":"ide1-cd1","force":true}}
Feb 16 16:04:19 localhost xenopsd-xc: [debug|localhost.localdomain|7 ||xenops] Got QMP event, domain-4: DEVICE_TRAY_MOVED
Feb 16 16:04:19 localhost xenopsd-xc: [ info|localhost.localdomain|41 |Async.VBD.eject R:9a2c66e44a71|xenops] File "xc/device_common.ml", line 384, characters 10-17: skipping unexpected qmp response from domid 4: {"event":"DEVICE_TRAY_MOVED","timestamp":{"seconds":1518797059,"microseconds":625362}}

```

Reply:
If those unexpected messages are completely normal and harmless, should they be downgraded to debug?